metrics/stackdriver_exporter.go

```diff
@@ -198,45 +198,51 @@ func newStackdriverExporter(config *metricsConfig, logger *zap.SugaredLogger) (v
 		nil
 }
 
-func sdCustomMetricsRecorder(mc metricsConfig) func(context.Context, []stats.Measurement, ...stats.Options) error {
+func sdCustomMetricsRecorder(mc metricsConfig, allowCustomMetrics bool) func(context.Context, []stats.Measurement, ...stats.Options) error {
 	gm := getMergedGCPMetadata(&mc)
 	metadataMap := map[string]string{
 		metricskey.LabelProject:     gm.project,
 		metricskey.LabelLocation:    gm.location,
 		metricskey.LabelClusterName: gm.cluster,
 	}
 	return func(ctx context.Context, mss []stats.Measurement, ros ...stats.Options) error {
-		// Filter the measuremenst array to only include permitted metrics.
 		i := 0
 		var templ *resourceTemplate
+		templateFound := false
+
+		// This loop serves two purpose, filtering out custom metrics when allowCustomMetrics set to false,
+		// and find the template for the system metrics.
 		for _, m := range mss {
 			metricType := path.Join(mc.stackdriverMetricTypePrefix, m.Measure().Name())
-			if t, ok := metricToResourceLabels[metricType]; ok {
+			t, ok := metricToResourceLabels[metricType]
+			if ok || allowCustomMetrics {
 				mss[i] = m
 				i++
-				if templ != nil && templ != t {
+				if templateFound && templ != t {
```

Comment:
It looks like introducing a limitation that `metrics.RecordBatch` can only accept measurements with the same resource type. @evankanderson and @jjzeng-seattle could you confirm this? If so, maybe add a comment to `RecordBatch` to mention it.

Reply:
Yes, that is the intent. I don't think this has an effect on practice, but I'll double-check.

I'll add a comment.

Reply:
If this becomes an issue, we could split the record into several Record calls, but I'd prefer to do that in a followup PR.

Reply:
I don't think it's an issue.  In the current code, all the measurements in `RecordBatch` use the same `context`, which means they have the same metric tags.

Reply:
Actually, it turns out there is one usage in KPA that runs afoul of this today:

```golang
	stats := []stats.Measurement{
		actualPodCountM.M(int64(pc.ready)), notReadyPodCountM.M(int64(pc.notReady)),
		pendingPodCountM.M(int64(pc.pending)), terminatingPodCountM.M(int64(pc.terminating)),
	}
	// Negative "want" values represent an empty metrics pipeline and thus no specific request is being made.
	if pc.want >= 0 {
		stats = append(stats, requestedPodCountM.M(int64(pc.want)))
	}
	pkgmetrics.RecordBatch(ctx, stats...)
```

https://github.com/knative/serving/blob/master/pkg/reconciler/autoscaling/kpa/kpa.go#L220

I'd still prefer to fix this by a followup PR, either to serving or to the stackdriver exporter, if that's okay. (I'm not sure which makes more sense -- it seems like this only affects reporting `knative.dev/serving/autoscaler/requested_pods`, `knative.dev/serving/autoscaler/not_ready_pods` and `knative.dev/serving/autoscaler/terminating_pods` when allowCustomMetrics is true.)